src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -303,8 +303,15 @@ ReadFromMergeTree::ReadFromMergeTree(
 {
     if (is_parallel_reading_from_replicas)
     {
-        all_ranges_callback = all_ranges_callback_.value_or(context->getMergeTreeAllRangesCallback());
-        read_task_callback = read_task_callback_.value_or(context->getMergeTreeReadTaskCallback());
+        if (all_ranges_callback_.has_value())
+            all_ranges_callback = all_ranges_callback_.value();
+        else
+            all_ranges_callback = context->getMergeTreeAllRangesCallback();
```

Comment:
is it what `value_or` supposed to do?

Reply:
Yes, but it's usage lead to https://github.com/ClickHouse/ClickHouse/pull/64448#issuecomment-2240264753